### ISSUE TITLE
how  to use labels or text as conditions

### ISSUE THREAD

**Issue by zhuyyx** [xandergos/sCM-mnist#1]
Hi, great work!
I'm also trying to reproduce sCM and testing with your code.
"""
train_consistency.py  line 90~98
v_x = torch.cos(t) * torch.sin(t) * dxt_dt / sigma_data
v_t = torch.cos(t) * torch.sin(t)
F_theta, F_theta_grad, logvar = torch.func.jvp(
    model_wrapper, 
    (x_t / sigma_data, t),
    (v_x, v_t),
    has_aux=True
)
"""

I have a question: if I want to use labels or text as conditions, what should v_y look like?
I'm looking forward to your answer.

**Comment by xandergos** [comment]
I assume $y$ refers to the conditional input to the model. In that case, there is no $v_y$. The model only needs to be differentiated with respect to $x$ and $t$, while $y$ is just treated as a constant. With class labels, you can think of this like training a separate model for each class, but share most of the weights to improve performance.

`model_wrapper` exists so that only the necessary variables are differentiated: `scaled_x_t` and `t`. `return_logvar` is not (and can't be) differentiated, so it is not used as input to model_wrapper:

```python
def model_wrapper(scaled_x_t, t):
        pred, logvar = model(scaled_x_t, t.flatten(), return_logvar=True)
        return pred, logvar
```

If you have a separate condition `c` to the model (class label, text, etc), you can modify `model_wrapper` to capture `c` and input it into the model:
```python
for x0, class_label in dataset:
    x0 = x0.to(device)
    class_label = class_label.to(device)

    # Notice that c is captured as input to the model, but is not a parameter 
    # in model_wrapper, so it will not be differentiated (i.e it is treated as a constant).
    def model_wrapper(scaled_x_t, t):
        pred, logvar = model(scaled_x_t, t.flatten(), return_logvar=True, c=class_label)
        return pred, logvar
```

Keep in mind that you have define `model_wrapper` *after* loading each batch of the dataset with this modification.

I added a commit efc165954f7997ad246d05b4f3eaf15c9020e7fb to make this clearer.

**Comment by zhuyyx** [comment]
ok, understood, thank you very much.

**Comment by zhuyyx** [comment]
closed issues

**Comment by zhuyyx** [comment]
+ fixed #24
